pkg/informers/informers.go

```diff
@@ -0,0 +1,179 @@
+// Copyright 2020 The prometheus-operator Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package informers
+
+import (
+	"github.com/pkg/errors"
+	"github.com/prometheus-operator/prometheus-operator/pkg/listwatch"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/tools/cache"
+)
+
+// InformLister is the interface that both exposes a shared index informer
+// and a generic lister.
+// Usually generated clients declare this interface as "GenericInformer".
+type InformLister interface {
+	Informer() cache.SharedIndexInformer
+	Lister() cache.GenericLister
+}
+
+// FactoriesForNamespaces is a way to combine several shared informers into a single struct with unified listing power.
+type FactoriesForNamespaces interface {
+	ForResource(namespace string, resource schema.GroupVersionResource) (InformLister, error)
+	Namespaces() sets.String
+}
+
+// ForResource contains a slice of InformLister for a concrete resource type,
+// one per namespace.
+type ForResource struct {
```

Comment:
Why are we creating a struct with an field that is a slice of InformListers rather than directly aliasing the `[]InformLister` type, e.g.

```golang
type ForResource []InformLister
```

If we were adding and removing elements from the slice on-the-fly, then I agree we would want to use a struct in order to keep a consistent reference for the user, but it looks like this only occurs at creation-time.

Reply:
I have been thinking about this too but since this is not an algorithmic data structure but rather an infrastructural one I left it as a struct. In fact, as a follow-up I want to leave some logging debug traces where we want a logger in the struct too.